File: Reader.cpp
#include "Reader.h"

uint8_t Reader::peek_uint8(void)const noexcept{
	if(_s + 1 > _e){
		return 0;
	}
	return *_s;
}
uint32_t Reader::peek_uint(void)noexcept{
	uint32_t r = 0;
	auto s = _s;
	for(int i = 0; i < 35; i += 7){
		if(s + 1 > _e){
			break;
		}
		auto v = *s;
		++s;
		r |= (uint32_t)(v & 0x7F) << i;
		if(!(v & 0x80))break;
	}
	return r;
}
uint8_t Reader::read_uint8(void)noexcept{
	if(!_check(1))return 0;
	auto v = *_s;
	++_s;
	return v;
}
uint32_t Reader::read_uint32(void)noexcept{
	if(!_check(4))return 0;
	auto v = *(uint32_t *)_s;
	_s += 4;
	return v;
}
float Reader::read_float(void)noexcept{
	if(!_check(4))return 0;
	auto v = *(float *)_s;
	_s += 4;
	return v;
}
uint32_t Reader::read_uint(void)noexcept{
	uint32_t r = 0;
	for(int i = 0; i < 35; i += 7){
		auto v = read_uint8();
		r |= (uint32_t)(v & 0x7F) << i;
		if(!(v & 0x80))break;
	}
	return r;
}
const uint8_t *Reader::read_data(size_t size)noexcept{
	if(!_check(size))return nullptr;
	auto data = _s;
	_s += size;
	return data;
}

bool Reader::_check(size_t size)noexcept{
	if(_s + size > _e){
		_s = _e + 1;
		return false;
	}
	return true;
}

File: js/Reader.js
/**
 * Reader类 - 用于读取二进制数据
 */
class Reader {
    constructor(data) {
        if (data instanceof Uint8Array) {
            this._data = data;
        } else if (data instanceof ArrayBuffer) {
            this._data = new Uint8Array(data);
        } else {
            this._data = new Uint8Array(0);
        }
        this._ptr = 0;
        this._error = false;
    }

    get size() {
        return this._data.length;
    }

    get isEof() {
        return this._ptr >= this._data.length;
    }

    get isError() {
        return this._error;
    }

    setError() {
        this._error = true;
    }

    readUint8() {
        if (this._ptr >= this._data.length) {
            this.setError();
            return 0;
        }
        return this._data[this._ptr++];
    }

    readUint32() {
        if (this._ptr + 4 > this._data.length) {
            this.setError();
            return 0;
        }
        const view = new DataView(this._data.buffer, this._data.byteOffset + this._ptr, 4);
        this._ptr += 4;
        return view.getUint32(0, true); // little-endian
    }

    readFloat32() {
        if (this._ptr + 4 > this._data.length) {
            this.setError();
            return 0;
        }
        const view = new DataView(this._data.buffer, this._data.byteOffset + this._ptr, 4);
        this._ptr += 4;
        return view.getFloat32(0, true); // little-endian
    }

    readInt() {
        let result = 0n;
        for (let i = 0; i < 70; i += 7) {
            const v = this.readUint8();
            result |= BigInt(v & 0x7F) << BigInt(i);
            if (!(v & 0x80)) break;
        }
        // 转换为JavaScript Number（如果在安全范围内）
        if (result <= Number.MAX_SAFE_INTEGER) {
            return Number(result);
        }
        return result;
    }

    readData(size) {
        if (this._ptr + size > this._data.length) {
            this.setError();
            return new Reader(new Uint8Array(0));
        }
        const data = this._data.slice(this._ptr, this._ptr + size);
        this._ptr += size;
        return new Reader(data);
    }

    getData() {
        return this._data;
    }

    slice(start, end) {
        return this._data.slice(start, end);
    }
}

// 导出
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Reader;
}
